src/model/block/payload/transaction/output/mod.rs

```diff
@@ -142,13 +144,26 @@ pub enum Output {
 
 impl Output {
     /// Returns the [`Address`] that is in control of the output.
-    pub fn owning_address(&self) -> Option<&Address> {
+    pub fn owning_address(&self, spent_timestamp: impl Into<Option<MilestoneTimestamp>>) -> Option<&Address> {
         Some(match self {
             Self::Treasury(_) => return None,
             Self::Basic(BasicOutput {
                 address_unlock_condition,
+                expiration_unlock_condition,
                 ..
-            }) => &address_unlock_condition.address,
+            }) => {
+                if let (Some(spent_timestamp), Some(expiration_unlock_condition)) =
+                    (spent_timestamp.into(), expiration_unlock_condition)
+                {
+                    if spent_timestamp >= expiration_unlock_condition.timestamp {
+                        &expiration_unlock_condition.return_address
+                    } else {
+                        &address_unlock_condition.address
+                    }
+                } else {
+                    &address_unlock_condition.address
+                }
```

Comment:
have you considered factoring this out into a function to reduce code duplication?

Reply:
sure, why not